File: backend/app/routes/todos.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.todo import Todo
from app.schemas.todo import TodoCreate, TodoRead, TodoUpdate

router = APIRouter(prefix="/api/todos", tags=["todos"])


@router.get("", response_model=list[TodoRead])
def list_todos(db: Session = Depends(get_db)) -> list[TodoRead]:
    todos = db.scalars(select(Todo).order_by(Todo.id.desc())).all()
    return [TodoRead.model_validate(todo) for todo in todos]


@router.post("", response_model=TodoRead, status_code=status.HTTP_201_CREATED)
def create_todo(payload: TodoCreate, db: Session = Depends(get_db)) -> TodoRead:
    todo = Todo(title=payload.title.strip())
    db.add(todo)
    db.commit()
    db.refresh(todo)
    return TodoRead.model_validate(todo)


@router.patch("/{todo_id}", response_model=TodoRead)
def update_todo(todo_id: int, payload: TodoUpdate, db: Session = Depends(get_db)) -> TodoRead:
    todo = db.get(Todo, todo_id)
    if not todo:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Todo not found")

    data = payload.model_dump(exclude_unset=True)

    if "title" in data and data["title"] is not None:
        todo.title = data["title"].strip()
    if "completed" in data and data["completed"] is not None:
        todo.completed = data["completed"]

    db.add(todo)
    db.commit()
    db.refresh(todo)
    return TodoRead.model_validate(todo)


@router.delete("/{todo_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_todo(todo_id: int, db: Session = Depends(get_db)) -> None:
    todo = db.get(Todo, todo_id)
    if not todo:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Todo not found")

    db.delete(todo)
    db.commit()
    return None

File: frontend/src/services/todos.ts
import { api } from './api'
import type { Todo, TodoCreate, TodoUpdate } from '../types'

export async function fetchTodos(): Promise<Todo[]> {
  const response = await api.get<Todo[]>('/api/todos')
  return response.data
}

export async function createTodo(payload: TodoCreate): Promise<Todo> {
  const response = await api.post<Todo>('/api/todos', payload)
  return response.data
}

export async function updateTodo(id: number, payload: TodoUpdate): Promise<Todo> {
  const response = await api.patch<Todo>(`/api/todos/${id}`, payload)
  return response.data
}

export async function deleteTodo(id: number): Promise<void> {
  await api.delete(`/api/todos/${id}`)
}
